File: backend/src/main/java/com/backend/ord/config/RestClientConfig.java
package com.backend.ord.config;

import com.backend.ord.api.requests.openai.OpenAIRequest;
import com.backend.ord.api.responses.openai.OpenAIResponse;
import com.backend.ord.config.properties.OpenAIProperties;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.client.RestTemplate;

@Configuration
public class RestClientConfig {
    private final OpenAIProperties openAIProperties;

    public RestClientConfig(OpenAIProperties openAIProperties) {
        this.openAIProperties = openAIProperties;
    }

    public RestTemplate openAITemplate() {
        RestTemplate restTemplate = new RestTemplate();

        restTemplate.getInterceptors().add((request, body, execution) -> {
            request.getHeaders().set("Authorization", openAIProperties.getAuthenticationHeaderValue());
            return execution.execute(request, body);
        });

        return restTemplate;
    }

    public OpenAIResponse makeOpenAIPostRequest(OpenAIRequest request) {
        return openAITemplate().postForObject(
                openAIProperties.getApiUrl(),
                request,
                OpenAIResponse.class
        );
    }

}

File: backend/src/main/kotlin/com/backend/ord/config/RestClientConfig.kt
package com.backend.ord.config

import com.backend.ord.config.properties.OpenAIProperties
import com.backend.ord.core.ai_provider.dto.OpenAIRequest
import com.backend.ord.core.ai_provider.dto.OpenAIResponse
import com.backend.ord.exceptions.OpenAIResponseIsNullException
import com.backend.ord.utils.Console
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.web.client.RestTemplate

// TODO: Rename to the OpenAIClientConfig or similar
@Configuration
class RestClientConfig(
    private val openAIProperties: OpenAIProperties
) {
    private fun openAITemplate(): RestTemplate {
        val restTemplate = RestTemplate()

        restTemplate.interceptors.add(
            ClientHttpRequestInterceptor { request: HttpRequest,
                                           body: ByteArray,
                                           execution: ClientHttpRequestExecution ->
                request.headers["Authorization"] = openAIProperties.authenticationHeaderValue
                execution.execute(request, body)
            }
        )

        return restTemplate
    }

    fun makeOpenAIPostRequest(request: OpenAIRequest?): OpenAIResponse {
        Console.printPurple("Making OpenAI request...")

        return openAITemplate().postForObject(
            openAIProperties.apiUrl,
            request,
            OpenAIResponse::class.java
        ).let {
            it ?: throw OpenAIResponseIsNullException()
        }
    }
}
